src/pid.cpp

```diff
@@ -335,12 +336,11 @@ double Pid::computeCommand(double error, double error_dot, ros::Duration dt)
   // Calculate the integral of the position error
   i_error_ += dt.toSec() * p_error_;
 
-  if(gains.antiwindup_)
+  if(gains.antiwindup_ && gains.i_gain_!=0)
```

Comment:
The PR is ready for merge.

There's just one smaller item to discuss:
What should happen if `i_gain_` is 0 (e.g. temporarily)?
For now `i_error_`  would accumulate..
Beforehand, the behavior was undefined.
One option would be to set the error to 0 as well.

Reply:
That's a good question. Since `i_error_` would be clamped as soon as `i_gain_` is non-zero, I think the question is sort of like this: When `i_gain_` is changed from zero to non-zero, should `i_term` and `i_error_` be near the windup area (a. `i_error_` is accumulated when `i_gain_` is 0) or be near zero(b. `i_error_` is 0 when `i_gain_` is 0)? I think (a) is a good option too. What's your opinion?

Reply:
I am okay with option a, because - as you already pointed out - the error will be clamped as soon as the gain is non-zero.
That's what would happen with `antiwindup=false` anyway.
I am just not sure if the user expects this if the gain is moved from zero to non-zero for the first time.
We could even freeze the error while the gain is zero..

Reply:
In my opinion, the point is what the user expects when he sets `i_gain_` as zero. A PID controller whose `i_gain_` is zero (option a) or a PD controller (option b)? I agree that most users may expect option b. So how about changing the code as follows?

```diff
- i_error_ += dt.toSec() * p_error_;
+ if(gains.i_gain_==0)
+   i_error_ = 0;
+ else
+   i_error_ += dt.toSec() * p_error_;
```

Reply:
For now I'd like to keep the existing behavior for `antiwindup=false`.
If you agree to reset the error to zero, please do so inside  the `gains.antiwindup_` clause:
```c++
  if(gains.antiwindup_)
  {
    if (gains.i_gain_!=0)
    {
      // Prevent i_error_ from climbing higher than permitted by i_max_/i_min_
      boost::tuple<double, double> bounds = boost::minmax<double>(gains.i_min_ / gains.i_gain_, gains.i_max_ / gains.i_gain_);
      i_error_ = boost::algorithm::clamp(i_error_, bounds.get<0>(), bounds.get<1>());
    }
    else
    {
      // Reset i_error_
      i_error_ = 0
    }
  }
```
If you prefer option a, we can merge it as-is.
@bmagyar: Do you have any preference?

Reply:
I'd also prefer option a.